projects/igniteui-angular/src/lib/query-builder/query-builder-drag.service.ts

```diff
@@ -0,0 +1,568 @@
+            //have to wait a tick because upon blur, the next activeElement is always body, right before the next element gains focus
+            setTimeout(() => {
+                if (document.activeElement.className.indexOf(QueryBuilderSelectors.DRAG_INDICATOR) === -1) {
+                    this.resetDragAndDrop(true);
+                    this._keyboardSubscription$?.unsubscribe();
+                }
+            }, 0);
+        }
+    }
+
+    //Checks if the dragged ghost is horizontally on the same line with the drop ghost
+    private dragGhostIsOnDropGhostRow(dragGhost: HTMLElement, dropGhost: HTMLElement) {
+        const dragGhostBounds = dragGhost.getBoundingClientRect();
+        const dropGhostBounds = dropGhost.getBoundingClientRect();
+
+        if (!dragGhostBounds || !dropGhostBounds) return false;
+
+        const ghostHeight = dragGhostBounds.bottom - dragGhostBounds.top;
+
+        return !(dragGhostBounds.bottom < dropGhostBounds.top - ghostHeight || dragGhostBounds.top > dropGhostBounds.bottom + ghostHeight);
+    }
+
+    //Checks if the dragged ghost is north or south of a target element's center
+    private ghostInLowerPart(ofElement: HTMLElement) {
+        //if (event == null) return true;
+        const ghostBounds = this.dragGhostElement.getBoundingClientRect();
+        const targetBounds = ofElement.getBoundingClientRect();
+
+        return ((ghostBounds.top + ghostBounds.bottom) / 2) >= ((targetBounds.top + targetBounds.bottom) / 2);
+    }
+
+    //Create the drop ghost node based on the base chip that's been dragged
+    private createDropGhost(keyboardMode?: boolean) {
+        const dragCopy = this.sourceElement.cloneNode(true);
+        (dragCopy as HTMLElement).classList.add(QueryBuilderSelectors.FILTER_TREE_EXPRESSION_ITEM_DROP_GHOST);
+        (dragCopy as HTMLElement).style.display = '';
+        (dragCopy.firstChild as HTMLElement).style.visibility = 'visible';
+        dragCopy.removeChild(dragCopy.childNodes[3]);
+
+        if (!keyboardMode) {
+            var span = document.createElement('span')
+            span.innerHTML = this._queryBuilderTreeComponent.resourceStrings.igx_query_builder_drop_ghost_text;
+
+            dragCopy.firstChild.firstChild.removeChild(dragCopy.firstChild.firstChild.childNodes[1]);
+            dragCopy.firstChild.firstChild.removeChild(dragCopy.firstChild.firstChild.childNodes[1]);
+            (dragCopy.firstChild.firstChild.firstChild as HTMLElement).replaceChildren(span);
+            (dragCopy.firstChild.firstChild as HTMLElement).classList.add(QueryBuilderSelectors.FILTER_TREE_EXPRESSION_ITEM_GHOST);
+        }
+        return dragCopy;
+    }
+
+    //Make a copy of the drag chip and place it in the DOM north or south of the drop chip
+    private renderDropGhostChip(appendToElement: HTMLElement, appendUnder: boolean, keyboardMode?: boolean): void {
+        const dragCopy = this.createDropGhost(keyboardMode);
+
+        //Append the ghost
```

Comment:
See above. We want to minimize DOM manipulation in Angulara as much as possible. I am not going to stop the pr from being merged because of this but double check if it's possible to tempalte this please.